Add managarm as a tier 3 target

no92 opened the pull request:
This PR aims to introduce the `x86_64-unknown-managarm-mlibc` as a tier 3 target to Rust.

[managarm](https://github.com/managarm/managarm) is a microkernel with fully asynchronous I/O that also provides a POSIX server. Despite the differences, managarm provides good compatability with POSIX and Linux APIs. As a rule of thumb, barring OS-specific code, it should be mostly source-compatible with Linux.

We have been shipping a patched rust for over 25 releases now, and we would like to upstream our work. For a smoother process, this PR only adds the target to rustc and some documentation. `std` support will be added in a future PR.

## Addressing the tier 3 target policy

> A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

@no92, @64 and @Dennisbonke will be target maintainers.

> Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.
> - Target names should not introduce undue confusion or ambiguity unless absolutely necessary to maintain ecosystem compatibility. For example, if the name of the target makes people extremely likely to form incorrect beliefs about what it targets, the name should be changed or augmented to disambiguate it.
> - If possible, use only letters, numbers, dashes and underscores for the name. Periods (.) are known to cause issues in Cargo.

`x86_64-unknown-managarm-mlibc` is what we use for LLVM as well.

> Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.
> - The target must not introduce license incompatibilities.
> - Anything added to the Rust repository must be under the standard Rust license (MIT OR Apache-2.0).
> - The target must not cause the Rust tools or libraries built for any other host (even when supporting cross-compilation to the target) to depend on any new dependency less permissive than the Rust licensing policy. This applies whether the dependency is a Rust crate that would require adding new license exceptions (as specified by the tidy tool in the rust-lang/rust repository), or whether the dependency is a native library or binary. In other words, the introduction of the target must not cause a user installing or running a version of Rust or the Rust tools to be subject to any new license requirements.
> - Compiling, linking, and emitting functional binaries, libraries, or other code for the target (whether hosted on the target itself or cross-compiling from another target) must not depend on proprietary (non-FOSS) libraries. Host tools built for the target itself may depend on the ordinary runtime libraries supplied by the platform and commonly used by other applications built for the target, but those libraries must not be required for code generation for the target; cross-compilation to the target must not require such libraries at all. For instance, rustc built for the target may depend on a common proprietary C runtime library or console output library, but must not depend on a proprietary code generation library or code optimization library. Rust's license permits such combinations, but the Rust project has no interest in maintaining such combinations within the scope of Rust itself, even at tier 3.
> - "onerous" here is an intentionally subjective term. At a minimum, "onerous" legal/licensing terms include but are not limited to: non-disclosure requirements, non-compete requirements, contributor license agreements (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional on the employer or employment of any particular Rust developers, revocable terms, any requirements that create liability for the Rust project or its developers or users, or any requirements that adversely affect the livelihood or prospects of the Rust project or its developers or users.

[managarm](https://github.com/managarm/managarm) is licensed as MIT. No dependencies were added.

> Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.
> - This requirement does not prevent part or all of this policy from being cited in an explicit contract or work agreement (e.g. to implement or maintain support for a target). This requirement exists to ensure that a developer or team responsible for reviewing and approving a target does not face any legal threats or obligations that would prevent them from freely exercising their judgment in such approval, even if such judgment involves subjective matters or goes beyond the letter of these requirements.

Understood. None of the listed maintainers are on a Rust team.

> Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (core for most targets, alloc for targets that can support dynamic memory allocation, std for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.

Support for `std` will be provided in a future PR. Only minor changes are required, however they depend on support in the `libc` crate which will be PRed in soon.

> The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running binaries, or running tests (even if they do not pass), the documentation must explain how to run such binaries or tests for the target, using emulation if possible or dedicated hardware if necessary.

The steps needed to take are described in the documentation provided with this PR.

> Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.
> - Backlinks such as those generated by the issue/PR tracker when linking to an issue or PR are not considered a violation of this policy, within reason. However, such messages (even on a separate repository) must not generate notifications to anyone involved with a PR who has not requested such notifications.

Understood.

> Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.
> - In particular, this may come up when working on closely related targets, such as variations of the same architecture with different features. Avoid introducing unconditional uses of features that another variation of the target may not have; use conditional compilation or runtime detection, as appropriate, to let each target run code supported by that target.

We have no indication that anything breaks due to this PR.

> Tier 3 targets must be able to produce assembly using at least one of rustc's supported backends from any host target.

No problems here, as we target `x86_64`.

r? compiler-team

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8507328768/job/23298999510) [(plain)](https://github.com/rust-lang/rust/commit/bbe655b07f6af3cc5d974048a1749a297bcadee3/checks/23298999510/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v4' (SHA:5d5d22a31266ced268874388b861e4b58bb5c2f3)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.410   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.422 Collecting boolean-py==4.0
#13 0.424   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.437 Collecting chardet==5.1.0
---
#13 3.401 Building wheels for collected packages: reuse
#13 3.402   Building wheel for reuse (pyproject.toml): started
#13 3.727   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.728   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.729   Stored in directory: /tmp/pip-ephem-wheel-cache-sg41glxz/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.731 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.753   Attempting uninstall: setuptools
#13 3.753     Found existing installation: setuptools 59.6.0
#13 3.754     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#13 4.981   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#13 5.033      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 76.4 MB/s eta 0:00:00
#13 5.083 Collecting filelock<4,>=3.12.2
#13 5.086   Downloading filelock-3.13.3-py3-none-any.whl (11 kB)
#13 5.115 Collecting platformdirs<5,>=3.9.1
#13 5.118   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#13 5.135 Collecting distlib<1,>=0.3.7
#13 5.138   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.144      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 112.6 MB/s eta 0:00:00
#13 5.228 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.389 Successfully installed distlib-0.3.8 filelock-3.13.3 platformdirs-4.2.0 virtualenv-20.25.1
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      196544 kB
DirectMap2M:     8192000 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/871df0d13ae55fcfd8c880e99fedbaba35952e1c/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-871df0d13ae55fcfd8c880e99fedbaba35952e1c-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 26.39s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs at line 4:
     let mut base = base::managarm_mlibc::opts();
     base.cpu = "x86-64".to_string().into();
     base.max_atomic_width = Some(64);
-    base.pre_link_args.get_mut(&LinkerFlavor::Gnu(Cc::Yes, Lld::No)).unwrap().push("-m64".to_string().into());
+    base.pre_link_args
+        .get_mut(&LinkerFlavor::Gnu(Cc::Yes, Lld::No))
+        .unwrap()
+        .push("-m64".to_string().into());
     // don't use probe-stack=inline-asm until rust-lang/rust#83139 is resolved.
     base.stack_probes = StackProbeType::Call;
 
##[error]Diff in /checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs at line 12:
         llvm_target: "x86_64-unknown-managarm-mlibc".to_string().into(),
         pointer_width: 64,
         data_layout: "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
-            .to_string().into(),
+            .into(),
+            .into(),
         arch: "x86_64".to_string().into(),
         options: base,
-        metadata: TargetMetadata {
-            std: Some(false),
-            tier: Some(3),
-        }
-        }
+        metadata: TargetMetadata { std: Some(false), tier: Some(3), ..Default::default() },
 }
 
 
##[error]Diff in /checkout/compiler/rustc_target/src/spec/base/managarm_mlibc.rs at line 1:
-use crate::spec::{Cc, cvs, LinkArgs, LinkerFlavor, Lld, RelroLevel, TargetOptions};
+use crate::spec::{cvs, Cc, LinkArgs, LinkerFlavor, Lld, RelroLevel, TargetOptions};
 pub fn opts() -> TargetOptions {
     let mut args = LinkArgs::new();
     let mut args = LinkArgs::new();
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_teeos.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/hexagon_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv5te_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_sony_psx.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7neon_linux_androideabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7s_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_gnux32.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64ec_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_dragonfly.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_wrs_vxworks.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_nintendo_switch_freestanding.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_wrs_vxworks.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imac_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv6m_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_none_softfloat.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imafc_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/targets/s390x_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/csky_unknown_linux_gnuabiv2.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_pc_nto_qnx710.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7a_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64_linux_android.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_hurd_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_uwp_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_fuchsia.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa64r6el_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_kmc_solid_asp3.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imc_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_win7_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_muslabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7em_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_ios_macabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_gnueabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_haiku.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imac_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_linux_android.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/wasm32_wasip2.rs" "/checkout/compiler/rustc_target/src/spec/targets/msp430_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32ima_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i386_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32i_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_uclibceabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_pc_windows_gnullvm.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa32r6_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7m_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/wasm32_unknown_emscripten.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64e_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_unknown_linux_musl.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Mon Apr  1 11:14:17 UTC 2024
  network time: Mon, 01 Apr 2024 11:14:17 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8507435242/job/23299276736) [(plain)](https://github.com/rust-lang/rust/commit/5ef5bbb3207a56ec834a9178ec878d9269c10155/checks/23299276736/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v4' (SHA:5d5d22a31266ced268874388b861e4b58bb5c2f3)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.411   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.423 Collecting boolean-py==4.0
#13 0.426   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.437 Collecting chardet==5.1.0
---
#13 3.408 Building wheels for collected packages: reuse
#13 3.409   Building wheel for reuse (pyproject.toml): started
#13 3.734   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.735   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.735   Stored in directory: /tmp/pip-ephem-wheel-cache-jics81mh/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.738 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.759   Attempting uninstall: setuptools
#13 3.760     Found existing installation: setuptools 59.6.0
#13 3.761     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#13 4.993   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#13 5.061      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 57.6 MB/s eta 0:00:00
#13 5.112 Collecting filelock<4,>=3.12.2
#13 5.115   Downloading filelock-3.13.3-py3-none-any.whl (11 kB)
#13 5.142 Collecting platformdirs<5,>=3.9.1
#13 5.145   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#13 5.161 Collecting distlib<1,>=0.3.7
#13 5.165   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.171      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 109.2 MB/s eta 0:00:00
#13 5.257 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.418 Successfully installed distlib-0.3.8 filelock-3.13.3 platformdirs-4.2.0 virtualenv-20.25.1
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      192448 kB
DirectMap2M:     6098944 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/871df0d13ae55fcfd8c880e99fedbaba35952e1c/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-871df0d13ae55fcfd8c880e99fedbaba35952e1c-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 26.50s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs at line 14:
     Target {
         llvm_target: "x86_64-unknown-managarm-mlibc".to_string().into(),
         pointer_width: 64,
-        data_layout: "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
-            .into(),
+        data_layout:
+        data_layout:
+            "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+                .into(),
+                .into(),
         arch: "x86_64".to_string().into(),
         options: base,
         metadata: TargetMetadata { std: Some(false), tier: Some(3), ..Default::default() },
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_l4re_uclibc.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64le_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu_ilp32.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips64_openwrt_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_apple_darwin.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_uefi.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/nvptx64_nvidia_cuda.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7a_kmc_solid_asp3_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_uclibceabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv8m_base_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64imac_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armeb_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_linux_androideabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_teeos.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/hexagon_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv5te_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_sony_psx.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7neon_linux_androideabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7s_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_gnux32.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64ec_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_dragonfly.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_wrs_vxworks.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_nintendo_switch_freestanding.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_wrs_vxworks.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imac_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv6m_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_none_softfloat.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imafc_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/targets/s390x_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/csky_unknown_linux_gnuabiv2.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_pc_nto_qnx710.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7a_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64_linux_android.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_hurd_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_uwp_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_fuchsia.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa64r6el_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_kmc_solid_asp3.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imc_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_win7_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_muslabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7em_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_ios_macabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_gnueabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_haiku.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unikraft_linux_musl.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Mon Apr  1 11:25:26 UTC 2024
  network time: Mon, 01 Apr 2024 11:25:26 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8507489394/job/23299426507) [(plain)](https://github.com/rust-lang/rust/commit/8b2f75089b01fab3ac82db41fe83549ed3102cfd/checks/23299426507/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v4' (SHA:5d5d22a31266ced268874388b861e4b58bb5c2f3)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.422   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.434 Collecting boolean-py==4.0
#13 0.438   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.450 Collecting chardet==5.1.0
---
#13 3.453 Building wheels for collected packages: reuse
#13 3.454   Building wheel for reuse (pyproject.toml): started
#13 3.781   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.782   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.782   Stored in directory: /tmp/pip-ephem-wheel-cache-bohuqw5o/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.785 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.807   Attempting uninstall: setuptools
#13 3.808     Found existing installation: setuptools 59.6.0
#13 3.809     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.809     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.809     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.484 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.485 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.992 Collecting virtualenv
#13 5.033   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#13 5.084      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 76.7 MB/s eta 0:00:00
#13 5.120 Collecting distlib<1,>=0.3.7
#13 5.124   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.130      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 106.9 MB/s eta 0:00:00
#13 5.158 Collecting platformdirs<5,>=3.9.1
#13 5.161   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#13 5.193 Collecting filelock<4,>=3.12.2
#13 5.197   Downloading filelock-3.13.3-py3-none-any.whl (11 kB)
#13 5.283 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.446 Successfully installed distlib-0.3.8 filelock-3.13.3 platformdirs-4.2.0 virtualenv-20.25.1
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      214976 kB
DirectMap2M:     6076416 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/3d5528c287860b918e178a34f04ff903325571b3/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-3d5528c287860b918e178a34f04ff903325571b3-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 26.51s
##[endgroup]
fmt check
tidy check
tidy error: tests/assembly/targets/: missing assembly test for x86_64_unknown_managarm_mlibc
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 41.6 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 18.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 18.9 MB/s eta 0:00:00
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 128.2 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:01:03
  local time: Mon Apr  1 11:32:36 UTC 2024
  network time: Mon, 01 Apr 2024 11:32:36 GMT
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8507578819/job/23299660197) [(plain)](https://github.com/rust-lang/rust/commit/3f149bf69ea0a5d52c69514843f51ecce30ddcbd/checks/23299660197/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 30.0s done
#16 DONE 32.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.47s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
running 247 tests
........................................................................................  88/247
........................................................................................ 176/247
............................................................
spec::tests::x86_64_unknown_managarm_mlibc ... F

failures:


---- spec::tests::x86_64_unknown_managarm_mlibc stdout ----
thread 'spec::tests::x86_64_unknown_managarm_mlibc' panicked at compiler/rustc_target/src/spec/tests/tests_impl.rs:88:17:
assertion `left == right` failed
  left: Some(["-Wl,--as-needed", "-Wl,-z,noexecstack", "-m64"])
 right: None
error: test failed, to rerun pass `-p rustc_target --lib`


failures:
failures:
    spec::tests::x86_64_unknown_managarm_mlibc
test result: FAILED. 246 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 8.81ms

Build completed unsuccessfully in 0:28:40
  local time: Mon Apr  1 12:09:17 UTC 2024
```

</details>

Mark-Simulacrum commented on compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs:
```diff
@@ -0,0 +1,22 @@
+    base.cpu = "x86-64".to_string().into();
+    base.max_atomic_width = Some(64);
+    base.add_pre_link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &["-m64"]);
+    // don't use probe-stack=inline-asm until rust-lang/rust#83139 is resolved.
```
https://github.com/rust-lang/rust/issues/83139 is closed since 2021, should this point to something else? Just outdated?

Mark-Simulacrum commented on src/doc/rustc/src/platform-support/managarm.md:
```diff
@@ -0,0 +1,63 @@
+
+[build]
+target = ["x86_64-unknown-managarm-mlibc", "x86_64-unknown-linux-gnu"]
+build-dir = "/path/to/the/build/dir"
+docs = false
+
+[install]
+prefix = "/path/to/the/prefix"
+sysconfdir = "etc"
```
It seems likely these are all optional fields?

Mark-Simulacrum commented on src/doc/rustc/src/platform-support/managarm.md:
```diff
@@ -0,0 +1,63 @@
+sysconfdir = "etc"
+
+[rust]
+codegen-tests = false
```
Why not enable codegen-tests?

Mark-Simulacrum commented on src/doc/rustc/src/platform-support/managarm.md:
```diff
@@ -0,0 +1,63 @@
+
+[rust]
+codegen-tests = false
+deny-warnings = false
```
This should also be dropped, it's an optional field that's up to the user.

Mark-Simulacrum commented on src/doc/rustc/src/platform-support/managarm.md:
```diff
@@ -0,0 +1,63 @@
+
+## Building Rust programs
+
+Build a `x86_64-managarm-gcc` using our [gcc fork](https://github.com/managarm/gcc).
```
Does lld not work for this target?

no92 replied:
Using `lld` is not a supported and tested option for us right now.

wesleywiser commented on src/doc/rustc/src/platform-support/managarm.md:
```diff
@@ -0,0 +1,63 @@
+
+## Building the target
+
+For now, building a patched LLVM with [our patches located here](https://github.com/managarm/bootstrap-managarm/tree/master/patches/llvm) is necessary.
```
I don't think any other targets currently need a patched LLVM to build (of course, newer LLVMs might work _better_). If I remember correctly, the loongarch targets waited until we had updated to a version of LLVM with sufficient support to link `libcore` before being merged. 

@no92 is there urgency to land this PR now or can we wait until an LLVM release picks up support for managarm?

64 replied:
There's no urgency, so we're happy to merge our changes into LLVM first if that makes more sense.

But it would be nice if we could get this merged fairly soon after our changes hit LLVM master, because it blocks us upstreaming our changes to other Rust crates, and that's the bulk of the maintenance effort on our end: supporting locally patched crates (and sometimes multiple versions thereof!) has proved quite tricky to manage

bors commented:
:umbrella: The latest upstream changes (presumably #123517) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

davidtwco commented:
Marking this as blocked on LLVM support for the target

brad0 commented:
Just FYI Managarm support has been upstramed to LLVM.. https://github.com/llvm/llvm-project/commit/0f302f38b0014a0018031ffb3cb898fdc7d90880  and will be included with the LLVM 21 release.

no92 committed:
bootstrap: bump cc-rs to 1.2.28

no92 committed:
compiler: Add `{x86_64,aarch64,riscv64gc}-unknown-managarm-mlibc` targets

Co-authored-by: Dennis Bonke <dennis@managarm.org>

no92 committed:
doc: Add `*-unknown-managarm-mlibc` documentation

no92 commented:
Now that we're upstream in LLVM 21, we have updated our toolchain and rebased this patchset onto current rust master. Due to changes done to cc-rs in the meantime, we had to get the `mlibc` env recognized by it. Therefore, we need to bump the version of cc-rs that is used by bootstrap, as can be seen in the first commit. Also, we added aarch64 and riscv64 support, which we test and use on managarm already.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17308141678/job/49136246331) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49136246331) [(plain)](https://github.com/rust-lang/rust/commit/675c8caca63b92bc1dc30e0f8c2ae2575cfe313c/checks/49136246331/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 11.125
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs:1:
-use crate::spec::{base, Cc, LinkerFlavor, Lld, StackProbeType, Target};
+use crate::spec::{Cc, LinkerFlavor, Lld, StackProbeType, Target, base};
 
 pub(crate) fn target() -> Target {
     let mut base = base::managarm_mlibc::opts();
Diff in /checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_managarm_mlibc.rs:16:
             std: Some(false),
         },
         pointer_width: 64,
-        data_layout: "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
-            .into(),
+        data_layout:
+            "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128".into(),
         arch: "x86_64".into(),
-        options: base
+        options: base,
     }
 }
 
Diff in /checkout/compiler/rustc_target/src/spec/base/managarm_mlibc.rs:1:
-use crate::spec::{Cc, cvs, LinkArgs, LinkerFlavor, Lld, RelroLevel, TargetOptions};
+use crate::spec::{Cc, LinkArgs, LinkerFlavor, Lld, RelroLevel, TargetOptions, cvs};
 
 pub(crate) fn opts() -> TargetOptions {
     let mut args = LinkArgs::new();
Diff in /checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_managarm_mlibc.rs:1:
-use crate::spec::{base, StackProbeType, Target};
+use crate::spec::{StackProbeType, Target, base};
 
 pub(crate) fn target() -> Target {
     let mut base = base::managarm_mlibc::opts();
fmt: checked 6341 files
Bootstrap failed while executing `test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:47
  local time: Thu Aug 28 21:22:31 UTC 2025
  network time: Thu, 28 Aug 2025 21:22:31 GMT
```

</details>

bors commented:
:hourglass: Testing commit 8434d3810b938b2236180a67886cf79ee5d9efe4 with merge 846e377215c5985b3a0fe7af6bb71bdc45912aae...
<!-- homu: {"type":"BuildStarted","head_sha":"8434d3810b938b2236180a67886cf79ee5d9efe4","merge_sha":"846e377215c5985b3a0fe7af6bb71bdc45912aae"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/17341288994/job/49239353845)
Approved by: davidtwco
Pushing 846e377215c5985b3a0fe7af6bb71bdc45912aae to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/17341288994/job/49239353845"},"merge_sha":"846e377215c5985b3a0fe7af6bb71bdc45912aae"} -->

rust-timer commented:
Finished benchmarking commit (846e377215c5985b3a0fe7af6bb71bdc45912aae): [comparison URL](https://perf.rust-lang.org/compare.html?start=b53c72ffaaf10e17fef5deb063f2f3f3bc13c171&end=846e377215c5985b3a0fe7af6bb71bdc45912aae&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b53c72ffaaf10e17fef5deb063f2f3f3bc13c171&end=846e377215c5985b3a0fe7af6bb71bdc45912aae&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b53c72ffaaf10e17fef5deb063f2f3f3bc13c171&end=846e377215c5985b3a0fe7af6bb71bdc45912aae&stat=max-rss)
<details>
<summary>Results (primary -2.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.6% | [-2.6%, -2.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b53c72ffaaf10e17fef5deb063f2f3f3bc13c171&end=846e377215c5985b3a0fe7af6bb71bdc45912aae&stat=cycles:u)
<details>
<summary>Results (secondary 0.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [2.6%, 2.8%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.9% | [-4.9%, -4.9%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=b53c72ffaaf10e17fef5deb063f2f3f3bc13c171&end=846e377215c5985b3a0fe7af6bb71bdc45912aae&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 466.779s -> 465.855s (-0.20%)
**Artifact size**: 388.53 MiB -> 388.54 MiB (0.00%)
